src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs

```diff
@@ -288,6 +293,151 @@ bool tryBindAsHandlerType([NotNullWhen(true)] out BoundInterpolatedString? resul
+            {
+                int n = parts.Length - 1;
+                var formatString = PooledStringBuilder.GetInstance();
+                var stringBuilder = formatString.Builder;
+                var expressions = ArrayBuilder<BoundExpression>.GetInstance(n + 1);
+                expressions.Add(null!); // format placeholder
+                int nextFormatPosition = 0;
+                for (int i = 0; i <= n; i++)
+                {
+                    var part = parts[i];
+                    if (part is BoundStringInsert fillin)
+                    {
+                        // this is one of the expression holes
+                        stringBuilder.Append('{').Append(nextFormatPosition++);
+                        if (fillin.Alignment != null && !fillin.Alignment.HasErrors)
+                        {
+                            Debug.Assert(fillin.Alignment.ConstantValueOpt is { });
+                            stringBuilder.Append(',').Append(fillin.Alignment.ConstantValueOpt.Int64Value);
+                        }
+                        if (fillin.Format != null && !fillin.Format.HasErrors)
+                        {
+                            Debug.Assert(fillin.Format.ConstantValueOpt is { });
+                            stringBuilder.Append(':').Append(fillin.Format.ConstantValueOpt.StringValue);
+                        }
+                        stringBuilder.Append('}');
+                        var value = fillin.Value;
+                        if (value.Type?.TypeKind == TypeKind.Dynamic)
+                        {
+                            // Object type is checked by BindInterpolatedStringParts
+                            value = GenerateConversionForAssignment(Compilation.ObjectType, value, diagnostics);
+                        }
+
+                        expressions.Add(value); // NOTE: must still be lowered
+                    }
+                    else
+                    {
+                        Debug.Assert(part is BoundLiteral && part.ConstantValueOpt?.StringValue != null);
+                        // this is one of the literal parts.  If it contains a { or } then we need to escape those so that
+                        // they're treated the same way in string.Format.
+                        stringBuilder.Append(escapeInterpolatedStringLiteral(part.ConstantValueOpt.StringValue));
+                    }
+                }
+
+                expressions[0] = new BoundLiteral(syntax, ConstantValue.Create(formatString.ToStringAndFree()), GetSpecialType(Microsoft.CodeAnalysis.SpecialType.System_String, diagnostics, syntax)) { WasCompilerGenerated = true };
+                return expressions.ToImmutableAndFree();
+            }
+
+            static string escapeInterpolatedStringLiteral(string value)
+            {
+                var builder = PooledStringBuilder.GetInstance();
```

Comment:
I know you're just copying this over, but... this double builder seems useless to me. This method is only ever called to append to a different builder. Consider just passing in `stringBuilder` and making this return void.